arrabiata/README.md

```diff
@@ -0,0 +1,93 @@
+## Arrabiata - a generic recursive zero-knowledge argument implementation based on folding schemes
+
+### Motivation
+
+This library provides an implementation of a generic recursive zero-knowledge
+argument based on folding schemes (initially defined in
+[Nova](https://eprint.iacr.org/2021/370)), over the
+[pasta](https://github.com/zcash/pasta_curves) curves and using the IPA
+polynomial commitment.
+
+The end goal of this repository is to implement a Nova-like prover, and have a zkApp
+on the Mina blockchain verifying the recursive proof*. This way, any user can run
+arbitrarily large computation on their machine, make a proof, and rely on the
+SNARK workers to verify the proof is correct and include it on-chains.
+
+The first iteration of the project will allow to fold a polynomial-time function
+`f` (which can be, in a near future, a zkApp), of degree 2**. No generic lookup
+argument will be implemented in the first version, even though a "runtime"
+lookup/permutation argument will be required
+for cross-cells referencing. The implementation leverages different
+constructions and ideas described in papers like
+[ProtoGalaxy](https://eprint.iacr.org/2023/1106),
+[ProtoStar](https://eprint.iacr.org/2023/620), etc.
+
+*This might need changes to the Mina blockchain, with a possible new hardfork.
+Not even sure it is possible right now.
+
+**This will change. We might go up to degree 6 or 7, as we're building the
+different gadgets (EC addition, EC scalar multiplication, Poseidon).
+
+### Implementation details
+
+We provide a binary to run arbitrarily large computation.
+The implementation of the circuits will follow the one used by the o1vm
+interpreter. An interpreter will be implemented over an abstract environment.
+
+The environment used for the witness will contain all the information required
+to make an IVC proof, i.e. the current witness in addition to the oracle, the
+folding accumulator, etc. And the implementation attempts to have the smallest
+memory footprint it can have by optimising the use of the CPU cache and using
+references as much as it can. The implementation attempts to prioritize
+allocations on the stack when possible, and attempts to keep in the CPU cache as
+many data as required as long as possible to avoid indirection.
+
+The witness interpreter attempts to perform most of the algebraic operations
+using integers, and not field element. The results are reduced into the field
+when mandatory.
+
+While building the witness, the cross terms are also computed on the fly, to be
+used in the next iteration. This way, the prover only pays the price of the
```

Comment:
I don't get how the on the fly computation makes the prover only pay for the activated gate

Reply:
Because not all columns are used by all gates.

Reply:
My comment isn't correct. The witness is the evaluation, but the cross-terms require the coefficients of the polynomial.